File: 901.股票价格跨度.cpp
/*
 * @lc app=leetcode.cn id=901 lang=cpp
 *
 * [901] 股票价格跨度
 */

// @lc code=start

#include <stack>
#include <unordered_map>

class StockSpanner {
public:
    StockSpanner() {

    }
    
    int next(int price) {
        int pop_cnt = 0;
        while (!price_index.empty() && price >= price_index.top().first) {
            pop_cnt += price_index.top().second;
            price_index.pop();
        }
        price_index.push({price, pop_cnt + 1});
        return price_index.top().second;
    }

private:
    std::stack<std::pair<int, int>> price_index;
};

/**
 * Your StockSpanner object will be instantiated and called as such:
 * StockSpanner* obj = new StockSpanner();
 * int param_1 = obj->next(price);
 */
// @lc code=end

File: 901.股票价格跨度.rs
/*
 * @lc app=leetcode.cn id=901 lang=rust
 *
 * [901] 股票价格跨度
 */

// @lc code=start
struct StockSpanner {
    // price/count
    stack: Vec::<(i32, i32)>,
}


/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl StockSpanner {

    fn new() -> Self {
        StockSpanner { stack: vec![] }
    }
    
    fn next(&mut self, price: i32) -> i32 {
        let mut current = (price, 1);
        while !self.stack.is_empty() && self.stack[self.stack.len() - 1].0 <= current.0 {
            current.1 += self.stack[self.stack.len() - 1].1;
            self.stack.pop();
        }
        self.stack.push(current);
        current.1
    }
}

/**
 * Your StockSpanner object will be instantiated and called as such:
 * let obj = StockSpanner::new();
 * let ret_1: i32 = obj.next(price);
 */
// @lc code=end

